dialog_design.md

```diff
@@ -0,0 +1,68 @@
+# Dialogs
+
+Goal: Display a message and an optional graphic to all users, or some, or one, and return a response.
+
+## Architecture
+
+Create a new `Dialog` class that manages calls and responses. It is also a good place to park static convenience
+functions and event listeners. I propose a couple different setup options:
+
+1. `type` - What sort of response we should expect.
+	1. A `message` dialog presents an OK button, and returns no value, just a confirmation that OK was clicked.
+	2. A `confirm` dialog presents OK and Cancel buttons, and returns a boolean.
+	3. A `prompt` dialog presents a text field and OK/Cancel buttons, and returns the contents of the text field.
+2. `text` - The text that the dialog displays.
+3. `recipients` (optional) - The dialog is displayed to all currently connected users by default, but can be restricted
+	to a subset of users with this option. Takes either a `GroupMask`, a `User[]`, or just one `User`. This filtering
+	could be implemented in the sync layer fairly easily, I think.
+4. `icon` (optional) - A graphic to display alongside the text. Can be one of the built-in icons (`info`, `warning`,
+	`error`, or `question`), or a custom texture supplied by the developer.
+5. `persistent` (optional) - If `true`, the dialog is presented to users in the recipients list even if they are
+	not currently online. I imagine this would be implemented by an extra handler to `context.onUserJoined`, with
+	special logic to not send the dialog to the same user more than once. This is complicated by changing membership
+	in groups, not sure how to handle that yet.
+
+Once the dialog is set up, you would call `dialog.send()` to deliver the messages to clients. If you need to hide
```

Comment:
Let's rename this `show()`. Up until now we haven't emphasized that anything about this API is network-centric.

Reply:
It would be nice if `show()` returned a `Promise<DialogResult[]>` (see mention of the `DialogResult` type below).

```ts
    const results = await Dialog.Create(this.context, options).show();
```

Reply:
Returning a promise works if there was just one user getting the result but with multiple recipients, would that work?

---
In reply to: [320870236](https://github.com/Microsoft/mixed-reality-extension-sdk/pull/444#discussion_r320870236) [](ancestors = 320870236)

Reply:
Agreed, promise doesn't really work if there are multiple responses. I guess it could get wired up to the `finished` event though.

---
In reply to: [321511608](https://github.com/microsoft/mixed-reality-extension-sdk/pull/444#discussion_r321511608) [](ancestors = 321511608,320870236)